{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Untitled26.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU",
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "380286e962254235b4a49ca3becb9683": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "VBoxModel",
          "model_module_version": "1.5.0",
          "state": {
            "_view_name": "VBoxView",
            "_dom_classes": [
              "widget-interact"
            ],
            "_model_name": "VBoxModel",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.5.0",
            "box_style": "",
            "layout": "IPY_MODEL_11644cd5bb8a4b86a90c3ba7fca2a09d",
            "_model_module": "@jupyter-widgets/controls",
            "children": [
              "IPY_MODEL_00a9610e16fc49c0905d06888c44f6ec",
              "IPY_MODEL_de9f5e391d154db1b7f1129225b39cc0"
            ]
          }
        },
        "11644cd5bb8a4b86a90c3ba7fca2a09d": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "00a9610e16fc49c0905d06888c44f6ec": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "FileUploadModel",
          "model_module_version": "1.5.0",
          "state": {
            "_view_name": "FileUploadView",
            "_counter": 1,
            "style": "IPY_MODEL_3b680612e5e2453ca5bf42a8d7499ccf",
            "_dom_classes": [],
            "description": "Introduce FASTA file",
            "multiple": false,
            "_model_name": "FileUploadModel",
            "data": [
              null
            ],
            "button_style": "",
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "accept": "",
            "_view_count": null,
            "disabled": false,
            "_view_module_version": "1.5.0",
            "error": "",
            "description_tooltip": null,
            "metadata": [
              {
                "name": "uniprot-reviewed_yes+organism__Homo+sapiens+(Human)+[9606]_+pain.fasta",
                "type": "",
                "size": 241756,
                "lastModified": 1639419450500
              }
            ],
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_2282b02b0e5b4c9c8e49938586b0eb78",
            "icon": "upload"
          }
        },
        "de9f5e391d154db1b7f1129225b39cc0": {
          "model_module": "@jupyter-widgets/output",
          "model_name": "OutputModel",
          "model_module_version": "1.0.0",
          "state": {
            "_view_name": "OutputView",
            "msg_id": "",
            "_dom_classes": [],
            "_model_name": "OutputModel",
            "outputs": [],
            "_view_module": "@jupyter-widgets/output",
            "_model_module_version": "1.0.0",
            "_view_count": null,
            "_view_module_version": "1.0.0",
            "layout": "IPY_MODEL_a9adc18bdd27416498bf9cebe159c9de",
            "_model_module": "@jupyter-widgets/output"
          }
        },
        "3b680612e5e2453ca5bf42a8d7499ccf": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "ButtonStyleModel",
          "model_module_version": "1.5.0",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "ButtonStyleModel",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "button_color": null,
            "font_weight": "",
            "_view_module_version": "1.2.0",
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "2282b02b0e5b4c9c8e49938586b0eb78": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        },
        "a9adc18bdd27416498bf9cebe159c9de": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "model_module_version": "1.2.0",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        }
      }
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# DeepTFactor\n",
        "\n",
        "> Kim GB, Gao Y, Palsson BO, Lee SY. **DeepTFactor: A deep learning-based tool for the prediction of transcription factors**. *Proc Natl Acad Sci U S A.* 2021 Jan 12;118(2):e2021171118. doi: 10.1073/pnas.2021171118. PMID: 33372147; PMCID: PMC7812831.\n",
        "\n",
        "Transcription factors (TF) are sequence-specific DNA-binding proteins that modulate gene expression. Conventionally, TFs have been identified through sequence homology analysis with DNA-binding domains of already-characterized TFs. In this work, authors developed a Deep Learning method that predicts whether a protein is a TF or not, given only its amino-acid sequence."
      ],
      "metadata": {
        "id": "zDpIll8XD4_D"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## 1. Set up environment"
      ],
      "metadata": {
        "id": "-JeBkUSzD8MC"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "GIUBsgs0x5_M",
        "outputId": "7dd302f5-0eb2-4e49-b24c-727ed11c6565"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-13 20:14:42--  https://repo.continuum.io/archive/Anaconda3-5.1.0-Linux-x86_64.sh\n",
            "Resolving repo.continuum.io (repo.continuum.io)... 104.18.201.79, 104.18.200.79, 2606:4700::6812:c94f, ...\n",
            "Connecting to repo.continuum.io (repo.continuum.io)|104.18.201.79|:443... connected.\n",
            "HTTP request sent, awaiting response... 301 Moved Permanently\n",
            "Location: https://repo.anaconda.com/archive/Anaconda3-5.1.0-Linux-x86_64.sh [following]\n",
            "--2021-12-13 20:14:43--  https://repo.anaconda.com/archive/Anaconda3-5.1.0-Linux-x86_64.sh\n",
            "Resolving repo.anaconda.com (repo.anaconda.com)... 104.16.130.3, 104.16.131.3, 2606:4700::6810:8203, ...\n",
            "Connecting to repo.anaconda.com (repo.anaconda.com)|104.16.130.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 577996269 (551M) [application/x-sh]\n",
            "Saving to: ‘Anaconda3-5.1.0-Linux-x86_64.sh’\n",
            "\n",
            "Anaconda3-5.1.0-Lin 100%[===================>] 551.22M  98.8MB/s    in 5.9s    \n",
            "\n",
            "2021-12-13 20:14:49 (94.0 MB/s) - ‘Anaconda3-5.1.0-Linux-x86_64.sh’ saved [577996269/577996269]\n",
            "\n",
            "PREFIX=/usr/local\n",
            "installing: python-3.6.4-hc3d631a_1 ...\n",
            "Python 3.6.4 :: Anaconda, Inc.\n",
            "installing: ca-certificates-2017.08.26-h1d4fec5_0 ...\n",
            "installing: conda-env-2.6.0-h36134e3_1 ...\n",
            "installing: intel-openmp-2018.0.0-hc7b2577_8 ...\n",
            "installing: libgcc-ng-7.2.0-h7cc24e2_2 ...\n",
            "installing: libgfortran-ng-7.2.0-h9f7466a_2 ...\n",
            "installing: libstdcxx-ng-7.2.0-h7a57d05_2 ...\n",
            "installing: bzip2-1.0.6-h9a117a8_4 ...\n",
            "installing: expat-2.2.5-he0dffb1_0 ...\n",
            "installing: gmp-6.1.2-h6c8ec71_1 ...\n",
            "installing: graphite2-1.3.10-hf63cedd_1 ...\n",
            "installing: icu-58.2-h9c2bf20_1 ...\n",
            "installing: jbig-2.1-hdba287a_0 ...\n",
            "installing: jpeg-9b-h024ee3a_2 ...\n",
            "installing: libffi-3.2.1-hd88cf55_4 ...\n",
            "installing: libsodium-1.0.15-hf101ebd_0 ...\n",
            "installing: libtool-2.4.6-h544aabb_3 ...\n",
            "installing: libxcb-1.12-hcd93eb1_4 ...\n",
            "installing: lzo-2.10-h49e0be7_2 ...\n",
            "installing: mkl-2018.0.1-h19d6760_4 ...\n",
            "installing: ncurses-6.0-h9df7e31_2 ...\n",
            "installing: openssl-1.0.2n-hb7f436b_0 ...\n",
            "installing: patchelf-0.9-hf79760b_2 ...\n",
            "installing: pcre-8.41-hc27e229_1 ...\n",
            "installing: pixman-0.34.0-hceecf20_3 ...\n",
            "installing: tk-8.6.7-hc745277_3 ...\n",
            "installing: unixodbc-2.3.4-hc36303a_1 ...\n",
            "installing: xz-5.2.3-h55aa19d_2 ...\n",
            "installing: yaml-0.1.7-had09818_2 ...\n",
            "installing: zlib-1.2.11-ha838bed_2 ...\n",
            "installing: glib-2.53.6-h5d9569c_2 ...\n",
            "installing: hdf5-1.10.1-h9caa474_1 ...\n",
            "installing: libedit-3.1-heed3624_0 ...\n",
            "installing: libpng-1.6.34-hb9fc6fc_0 ...\n",
            "installing: libssh2-1.8.0-h9cfc8f7_4 ...\n",
            "installing: libtiff-4.0.9-h28f6b97_0 ...\n",
            "installing: libxml2-2.9.7-h26e45fe_0 ...\n",
            "installing: mpfr-3.1.5-h11a74b3_2 ...\n",
            "installing: pandoc-1.19.2.1-hea2e7c5_1 ...\n",
            "installing: readline-7.0-ha6073c6_4 ...\n",
            "installing: zeromq-4.2.2-hbedb6e5_2 ...\n",
            "installing: dbus-1.12.2-hc3f9b76_1 ...\n",
            "installing: freetype-2.8-hab7d2ae_1 ...\n",
            "installing: gstreamer-1.12.4-hb53b477_0 ...\n",
            "installing: libcurl-7.58.0-h1ad7b7a_0 ...\n",
            "installing: libxslt-1.1.32-h1312cb7_0 ...\n",
            "installing: mpc-1.0.3-hec55b23_5 ...\n",
            "installing: sqlite-3.22.0-h1bed415_0 ...\n",
            "installing: curl-7.58.0-h84994c4_0 ...\n",
            "installing: fontconfig-2.12.4-h88586e7_1 ...\n",
            "installing: gst-plugins-base-1.12.4-h33fb286_0 ...\n",
            "installing: alabaster-0.7.10-py36h306e16b_0 ...\n",
            "installing: asn1crypto-0.24.0-py36_0 ...\n",
            "installing: attrs-17.4.0-py36_0 ...\n",
            "installing: backports-1.0-py36hfa02d7e_1 ...\n",
            "installing: beautifulsoup4-4.6.0-py36h49b8c8c_1 ...\n",
            "installing: bitarray-0.8.1-py36h14c3975_1 ...\n",
            "installing: boto-2.48.0-py36h6e4cd66_1 ...\n",
            "installing: cairo-1.14.12-h77bcde2_0 ...\n",
            "installing: certifi-2018.1.18-py36_0 ...\n",
            "installing: chardet-3.0.4-py36h0f667ec_1 ...\n",
            "installing: click-6.7-py36h5253387_0 ...\n",
            "installing: cloudpickle-0.5.2-py36_1 ...\n",
            "installing: colorama-0.3.9-py36h489cec4_0 ...\n",
            "installing: contextlib2-0.5.5-py36h6c84a62_0 ...\n",
            "installing: dask-core-0.16.1-py36_0 ...\n",
            "installing: decorator-4.2.1-py36_0 ...\n",
            "installing: docutils-0.14-py36hb0f60f5_0 ...\n",
            "installing: entrypoints-0.2.3-py36h1aec115_2 ...\n",
            "installing: et_xmlfile-1.0.1-py36hd6bccc3_0 ...\n",
            "installing: fastcache-1.0.2-py36h14c3975_2 ...\n",
            "installing: filelock-2.0.13-py36h646ffb5_0 ...\n",
            "installing: glob2-0.6-py36he249c77_0 ...\n",
            "installing: gmpy2-2.0.8-py36hc8893dd_2 ...\n",
            "installing: greenlet-0.4.12-py36h2d503a6_0 ...\n",
            "installing: heapdict-1.0.0-py36_2 ...\n",
            "installing: idna-2.6-py36h82fb2a8_1 ...\n",
            "installing: imagesize-0.7.1-py36h52d8127_0 ...\n",
            "installing: ipython_genutils-0.2.0-py36hb52b0d5_0 ...\n",
            "installing: itsdangerous-0.24-py36h93cc618_1 ...\n",
            "installing: jdcal-1.3-py36h4c697fb_0 ...\n",
            "installing: lazy-object-proxy-1.3.1-py36h10fcdad_0 ...\n",
            "installing: llvmlite-0.21.0-py36ha241eea_0 ...\n",
            "installing: locket-0.2.0-py36h787c0ad_1 ...\n",
            "installing: lxml-4.1.1-py36hf71bdeb_1 ...\n",
            "installing: markupsafe-1.0-py36hd9260cd_1 ...\n",
            "installing: mccabe-0.6.1-py36h5ad9710_1 ...\n",
            "installing: mistune-0.8.3-py36_0 ...\n",
            "installing: mkl-service-1.1.2-py36h17a0993_4 ...\n",
            "installing: mpmath-1.0.0-py36hfeacd6b_2 ...\n",
            "installing: msgpack-python-0.5.1-py36h6bb024c_0 ...\n",
            "installing: multipledispatch-0.4.9-py36h41da3fb_0 ...\n",
            "installing: numpy-1.14.0-py36h3dfced4_1 ...\n",
            "installing: olefile-0.45.1-py36_0 ...\n",
            "installing: pandocfilters-1.4.2-py36ha6701b7_1 ...\n",
            "installing: parso-0.1.1-py36h35f843b_0 ...\n",
            "installing: path.py-10.5-py36h55ceabb_0 ...\n",
            "installing: pep8-1.7.1-py36_0 ...\n",
            "installing: pickleshare-0.7.4-py36h63277f8_0 ...\n",
            "installing: pkginfo-1.4.1-py36h215d178_1 ...\n",
            "installing: pluggy-0.6.0-py36hb689045_0 ...\n",
            "installing: ply-3.10-py36hed35086_0 ...\n",
            "installing: psutil-5.4.3-py36h14c3975_0 ...\n",
            "installing: ptyprocess-0.5.2-py36h69acd42_0 ...\n",
            "installing: py-1.5.2-py36h29bf505_0 ...\n",
            "installing: pycodestyle-2.3.1-py36hf609f19_0 ...\n",
            "installing: pycosat-0.6.3-py36h0a5515d_0 ...\n",
            "installing: pycparser-2.18-py36hf9f622e_1 ...\n",
            "installing: pycrypto-2.6.1-py36h14c3975_7 ...\n",
            "installing: pycurl-7.43.0.1-py36hb7f436b_0 ...\n",
            "installing: pyodbc-4.0.22-py36hf484d3e_0 ...\n",
            "installing: pyparsing-2.2.0-py36hee85983_1 ...\n",
            "installing: pysocks-1.6.7-py36hd97a5b1_1 ...\n",
            "installing: pytz-2017.3-py36h63b9c63_0 ...\n",
            "installing: pyyaml-3.12-py36hafb9ca4_1 ...\n",
            "installing: pyzmq-16.0.3-py36he2533c7_0 ...\n",
            "installing: qt-5.6.2-h974d657_12 ...\n",
            "installing: qtpy-1.3.1-py36h3691cc8_0 ...\n",
            "installing: rope-0.10.7-py36h147e2ec_0 ...\n",
            "installing: ruamel_yaml-0.15.35-py36h14c3975_1 ...\n",
            "installing: send2trash-1.4.2-py36_0 ...\n",
            "installing: simplegeneric-0.8.1-py36_2 ...\n",
            "installing: sip-4.18.1-py36h51ed4ed_2 ...\n",
            "installing: six-1.11.0-py36h372c433_1 ...\n",
            "installing: snowballstemmer-1.2.1-py36h6febd40_0 ...\n",
            "installing: sortedcontainers-1.5.9-py36_0 ...\n",
            "installing: sphinxcontrib-1.0-py36h6d0f590_1 ...\n",
            "installing: sqlalchemy-1.2.1-py36h14c3975_0 ...\n",
            "installing: tblib-1.3.2-py36h34cf8b6_0 ...\n",
            "installing: testpath-0.3.1-py36h8cadb63_0 ...\n",
            "installing: toolz-0.9.0-py36_0 ...\n",
            "installing: tornado-4.5.3-py36_0 ...\n",
            "installing: typing-3.6.2-py36h7da032a_0 ...\n",
            "installing: unicodecsv-0.14.1-py36ha668878_0 ...\n",
            "installing: wcwidth-0.1.7-py36hdf4376a_0 ...\n",
            "installing: webencodings-0.5.1-py36h800622e_1 ...\n",
            "installing: werkzeug-0.14.1-py36_0 ...\n",
            "installing: wrapt-1.10.11-py36h28b7045_0 ...\n",
            "installing: xlrd-1.1.0-py36h1db9f0c_1 ...\n",
            "installing: xlsxwriter-1.0.2-py36h3de1aca_0 ...\n",
            "installing: xlwt-1.3.0-py36h7b00a1f_0 ...\n",
            "installing: babel-2.5.3-py36_0 ...\n",
            "installing: backports.shutil_get_terminal_size-1.0.0-py36hfea85ff_2 ...\n",
            "installing: bottleneck-1.2.1-py36haac1ea0_0 ...\n",
            "installing: cffi-1.11.4-py36h9745a5d_0 ...\n",
            "installing: conda-verify-2.0.0-py36h98955d8_0 ...\n",
            "installing: cycler-0.10.0-py36h93f1223_0 ...\n",
            "installing: cytoolz-0.9.0-py36h14c3975_0 ...\n",
            "installing: h5py-2.7.1-py36h3585f63_0 ...\n",
            "installing: harfbuzz-1.7.4-hc5b324e_0 ...\n",
            "installing: html5lib-1.0.1-py36h2f9c1c0_0 ...\n",
            "installing: jedi-0.11.1-py36_0 ...\n",
            "installing: networkx-2.1-py36_0 ...\n",
            "installing: nltk-3.2.5-py36h7532b22_0 ...\n",
            "installing: numba-0.36.2-np114py36hc6662d5_0 ...\n",
            "installing: numexpr-2.6.4-py36hc4a3f9a_0 ...\n",
            "installing: openpyxl-2.4.10-py36_0 ...\n",
            "installing: packaging-16.8-py36ha668100_1 ...\n",
            "installing: partd-0.3.8-py36h36fd896_0 ...\n",
            "installing: pathlib2-2.3.0-py36h49efa8e_0 ...\n",
            "installing: pexpect-4.3.1-py36_0 ...\n",
            "installing: pillow-5.0.0-py36h3deb7b8_0 ...\n",
            "installing: pyqt-5.6.0-py36h0386399_5 ...\n",
            "installing: python-dateutil-2.6.1-py36h88d3b88_1 ...\n",
            "installing: pywavelets-0.5.2-py36he602eb0_0 ...\n",
            "installing: qtawesome-0.4.4-py36h609ed8c_0 ...\n",
            "installing: scipy-1.0.0-py36hbf646e7_0 ...\n",
            "installing: setuptools-38.4.0-py36_0 ...\n",
            "installing: singledispatch-3.4.0.3-py36h7a266c3_0 ...\n",
            "installing: sortedcollections-0.5.3-py36h3c761f9_0 ...\n",
            "installing: sphinxcontrib-websupport-1.0.1-py36hb5cb234_1 ...\n",
            "installing: sympy-1.1.1-py36hc6d1c1c_0 ...\n",
            "installing: terminado-0.8.1-py36_1 ...\n",
            "installing: traitlets-4.3.2-py36h674d592_0 ...\n",
            "installing: zict-0.1.3-py36h3a3bf81_0 ...\n",
            "installing: astroid-1.6.1-py36_0 ...\n",
            "installing: bleach-2.1.2-py36_0 ...\n",
            "installing: clyent-1.2.2-py36h7e57e65_1 ...\n",
            "installing: cryptography-2.1.4-py36hd09be54_0 ...\n",
            "installing: cython-0.27.3-py36h1860423_0 ...\n",
            "installing: datashape-0.5.4-py36h3ad6b5c_0 ...\n",
            "installing: distributed-1.20.2-py36_0 ...\n",
            "installing: get_terminal_size-1.0.0-haa9412d_0 ...\n",
            "installing: gevent-1.2.2-py36h2fe25dc_0 ...\n",
            "installing: imageio-2.2.0-py36he555465_0 ...\n",
            "installing: isort-4.2.15-py36had401c0_0 ...\n",
            "installing: jinja2-2.10-py36ha16c418_0 ...\n",
            "installing: jsonschema-2.6.0-py36h006f8b5_0 ...\n",
            "installing: jupyter_core-4.4.0-py36h7c827e3_0 ...\n",
            "installing: matplotlib-2.1.2-py36h0e671d2_0 ...\n",
            "installing: navigator-updater-0.1.0-py36h14770f7_0 ...\n",
            "installing: nose-1.3.7-py36hcdf7029_2 ...\n",
            "installing: pandas-0.22.0-py36hf484d3e_0 ...\n",
            "installing: pango-1.41.0-hd475d92_0 ...\n",
            "installing: patsy-0.5.0-py36_0 ...\n",
            "installing: pyflakes-1.6.0-py36h7bd6a15_0 ...\n",
            "installing: pygments-2.2.0-py36h0d3125c_0 ...\n",
            "installing: pytables-3.4.2-py36h3b5282a_2 ...\n",
            "installing: pytest-3.3.2-py36_0 ...\n",
            "installing: scikit-learn-0.19.1-py36h7aa7ec6_0 ...\n",
            "installing: wheel-0.30.0-py36hfd4bba0_1 ...\n",
            "installing: astropy-2.0.3-py36h14c3975_0 ...\n",
            "installing: bkcharts-0.2-py36h735825a_0 ...\n",
            "installing: bokeh-0.12.13-py36h2f9c1c0_0 ...\n",
            "installing: flask-0.12.2-py36hb24657c_0 ...\n",
            "installing: jupyter_client-5.2.2-py36_0 ...\n",
            "installing: nbformat-4.4.0-py36h31c9010_0 ...\n",
            "installing: pip-9.0.1-py36h6c6f9ce_4 ...\n",
            "installing: prompt_toolkit-1.0.15-py36h17d85b1_0 ...\n",
            "installing: pylint-1.8.2-py36_0 ...\n",
            "installing: pyopenssl-17.5.0-py36h20ba746_0 ...\n",
            "installing: statsmodels-0.8.0-py36h8533d0b_0 ...\n",
            "installing: dask-0.16.1-py36_0 ...\n",
            "installing: flask-cors-3.0.3-py36h2d857d3_0 ...\n",
            "installing: ipython-6.2.1-py36h88c514a_1 ...\n",
            "installing: nbconvert-5.3.1-py36hb41ffb7_0 ...\n",
            "installing: seaborn-0.8.1-py36hfad7ec4_0 ...\n",
            "installing: urllib3-1.22-py36hbe7ace6_0 ...\n",
            "installing: ipykernel-4.8.0-py36_0 ...\n",
            "installing: odo-0.5.1-py36h90ed295_0 ...\n",
            "installing: requests-2.18.4-py36he2e5f8d_1 ...\n",
            "installing: scikit-image-0.13.1-py36h14c3975_1 ...\n",
            "installing: anaconda-client-1.6.9-py36_0 ...\n",
            "installing: blaze-0.11.3-py36h4e06776_0 ...\n",
            "installing: jupyter_console-5.2.0-py36he59e554_1 ...\n",
            "installing: notebook-5.4.0-py36_0 ...\n",
            "installing: qtconsole-4.3.1-py36h8f73b5b_0 ...\n",
            "installing: sphinx-1.6.6-py36_0 ...\n",
            "installing: anaconda-project-0.8.2-py36h44fb852_0 ...\n",
            "installing: jupyterlab_launcher-0.10.2-py36_0 ...\n",
            "installing: numpydoc-0.7.0-py36h18f165f_0 ...\n",
            "installing: widgetsnbextension-3.1.0-py36_0 ...\n",
            "installing: anaconda-navigator-1.7.0-py36_0 ...\n",
            "installing: ipywidgets-7.1.1-py36_0 ...\n",
            "installing: jupyterlab-0.31.5-py36_0 ...\n",
            "installing: spyder-3.2.6-py36_0 ...\n",
            "installing: _ipyw_jlab_nb_ext_conf-0.1.0-py36he11e457_0 ...\n",
            "installing: jupyter-1.0.0-py36_4 ...\n",
            "installing: anaconda-5.1.0-py36_2 ...\n",
            "installing: conda-4.4.10-py36_0 ...\n",
            "installing: conda-build-3.4.1-py36_0 ...\n",
            "installation finished.\n",
            "WARNING:\n",
            "    You currently have a PYTHONPATH environment variable set. This may cause\n",
            "    unexpected behavior when running the Python interpreter in Anaconda3.\n",
            "    For best results, please verify that your PYTHONPATH only points to\n",
            "    directories of packages that are compatible with the Python interpreter\n",
            "    in Anaconda3: /usr/local\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs: \n",
            "    - conda\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    certifi-2021.10.8          |   py37h06a4308_0         154 KB\n",
            "    pyopenssl-19.0.0           |           py37_0          82 KB\n",
            "    libgcc-ng-9.1.0            |       hdf63c60_0         8.1 MB\n",
            "    zlib-1.2.11                |       h7b6447c_3         120 KB\n",
            "    requests-2.26.0            |     pyhd3eb1b0_0          59 KB\n",
            "    cffi-1.15.0                |   py37h7f8727e_0         224 KB\n",
            "    sqlite-3.33.0              |       h62c20be_0         2.0 MB\n",
            "    asn1crypto-1.4.0           |             py_0          77 KB\n",
            "    conda-4.11.0               |   py37h06a4308_0        16.9 MB\n",
            "    charset-normalizer-2.0.4   |     pyhd3eb1b0_0          33 KB\n",
            "    tk-8.6.11                  |       h1ccaba5_0         3.2 MB\n",
            "    cryptography-2.3.1         |   py37hc365091_0         585 KB\n",
            "    pycosat-0.6.3              |   py37h27cfd23_0         108 KB\n",
            "    xz-5.2.5                   |       h7b6447c_0         438 KB\n",
            "    conda-package-handling-1.7.3|   py37h27cfd23_1         962 KB\n",
            "    openssl-1.0.2u             |       h7b6447c_0         3.1 MB\n",
            "    ncurses-6.3                |       h7f8727e_2         1.0 MB\n",
            "    pip-21.2.2                 |   py37h06a4308_0         2.0 MB\n",
            "    python-3.7.0               |       hc3d631a_0        31.7 MB\n",
            "    pysocks-1.7.1              |           py37_1          27 KB\n",
            "    _libgcc_mutex-0.1          |             main           3 KB\n",
            "    readline-7.0               |       h7b6447c_5         392 KB\n",
            "    ruamel_yaml-0.15.100       |   py37h27cfd23_0         267 KB\n",
            "    wheel-0.37.0               |     pyhd3eb1b0_1          31 KB\n",
            "    setuptools-58.0.4          |   py37h06a4308_0         979 KB\n",
            "    pycparser-2.21             |     pyhd3eb1b0_0          94 KB\n",
            "    libedit-3.1.20210910       |       h7f8727e_0         191 KB\n",
            "    six-1.16.0                 |     pyhd3eb1b0_0          18 KB\n",
            "    yaml-0.2.5                 |       h7b6447c_0          87 KB\n",
            "    tqdm-4.62.3                |     pyhd3eb1b0_1          79 KB\n",
            "    urllib3-1.26.7             |     pyhd3eb1b0_0         104 KB\n",
            "    brotlipy-0.7.0             |py37h27cfd23_1003         350 KB\n",
            "    ca-certificates-2021.10.26 |       h06a4308_2         121 KB\n",
            "    idna-3.3                   |     pyhd3eb1b0_0          55 KB\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:        73.7 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "    _libgcc_mutex:          0.1-main               \n",
            "    brotlipy:               0.7.0-py37h27cfd23_1003\n",
            "    charset-normalizer:     2.0.4-pyhd3eb1b0_0     \n",
            "    conda-package-handling: 1.7.3-py37h27cfd23_1   \n",
            "    tqdm:                   4.62.3-pyhd3eb1b0_1    \n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "    asn1crypto:             0.24.0-py36_0           --> 1.4.0-py_0              \n",
            "    ca-certificates:        2017.08.26-h1d4fec5_0   --> 2021.10.26-h06a4308_2   \n",
            "    certifi:                2018.1.18-py36_0        --> 2021.10.8-py37h06a4308_0\n",
            "    cffi:                   1.11.4-py36h9745a5d_0   --> 1.15.0-py37h7f8727e_0   \n",
            "    conda:                  4.4.10-py36_0           --> 4.11.0-py37h06a4308_0   \n",
            "    cryptography:           2.1.4-py36hd09be54_0    --> 2.3.1-py37hc365091_0    \n",
            "    idna:                   2.6-py36h82fb2a8_1      --> 3.3-pyhd3eb1b0_0        \n",
            "    libedit:                3.1-heed3624_0          --> 3.1.20210910-h7f8727e_0 \n",
            "    libgcc-ng:              7.2.0-h7cc24e2_2        --> 9.1.0-hdf63c60_0        \n",
            "    ncurses:                6.0-h9df7e31_2          --> 6.3-h7f8727e_2          \n",
            "    openssl:                1.0.2n-hb7f436b_0       --> 1.0.2u-h7b6447c_0       \n",
            "    pip:                    9.0.1-py36h6c6f9ce_4    --> 21.2.2-py37h06a4308_0   \n",
            "    pycosat:                0.6.3-py36h0a5515d_0    --> 0.6.3-py37h27cfd23_0    \n",
            "    pycparser:              2.18-py36hf9f622e_1     --> 2.21-pyhd3eb1b0_0       \n",
            "    pyopenssl:              17.5.0-py36h20ba746_0   --> 19.0.0-py37_0           \n",
            "    pysocks:                1.6.7-py36hd97a5b1_1    --> 1.7.1-py37_1            \n",
            "    python:                 3.6.4-hc3d631a_1        --> 3.7.0-hc3d631a_0        \n",
            "    readline:               7.0-ha6073c6_4          --> 7.0-h7b6447c_5          \n",
            "    requests:               2.18.4-py36he2e5f8d_1   --> 2.26.0-pyhd3eb1b0_0     \n",
            "    ruamel_yaml:            0.15.35-py36h14c3975_1  --> 0.15.100-py37h27cfd23_0 \n",
            "    setuptools:             38.4.0-py36_0           --> 58.0.4-py37h06a4308_0   \n",
            "    six:                    1.11.0-py36h372c433_1   --> 1.16.0-pyhd3eb1b0_0     \n",
            "    sqlite:                 3.22.0-h1bed415_0       --> 3.33.0-h62c20be_0       \n",
            "    tk:                     8.6.7-hc745277_3        --> 8.6.11-h1ccaba5_0       \n",
            "    urllib3:                1.22-py36hbe7ace6_0     --> 1.26.7-pyhd3eb1b0_0     \n",
            "    wheel:                  0.30.0-py36hfd4bba0_1   --> 0.37.0-pyhd3eb1b0_1     \n",
            "    xz:                     5.2.3-h55aa19d_2        --> 5.2.5-h7b6447c_0        \n",
            "    yaml:                   0.1.7-had09818_2        --> 0.2.5-h7b6447c_0        \n",
            "    zlib:                   1.2.11-ha838bed_2       --> 1.2.11-h7b6447c_3       \n",
            "\n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n",
            "no change     /usr/local/condabin/conda\n",
            "no change     /usr/local/bin/conda\n",
            "no change     /usr/local/bin/conda-env\n",
            "no change     /usr/local/bin/activate\n",
            "no change     /usr/local/bin/deactivate\n",
            "no change     /usr/local/etc/profile.d/conda.sh\n",
            "no change     /usr/local/etc/fish/conf.d/conda.fish\n",
            "no change     /usr/local/shell/condabin/Conda.psm1\n",
            "no change     /usr/local/shell/condabin/conda-hook.ps1\n",
            "no change     /usr/local/lib/python3.7/site-packages/xontrib/conda.xsh\n",
            "no change     /usr/local/etc/profile.d/conda.csh\n",
            "modified      /root/.bashrc\n",
            "\n",
            "==> For changes to take effect, close and re-open your current shell. <==\n",
            "\n",
            "Cloning into 'deeptfactor'...\n",
            "Unpacking objects: 100% (26/26), done.\n",
            "/content/deeptfactor\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Solving environment: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "ncurses-6.2          | 817 KB    | : 100% 1.0/1 [00:00<00:00,  1.86it/s]\n",
            "pickleshare-0.7.5    | 13 KB     | : 100% 1.0/1 [00:00<00:00,  7.31it/s]\n",
            "cudatoolkit-10.0.130 | 261.2 MB  | : 100% 1.0/1 [00:08<00:00,  8.20s/it]              \n",
            "jedi-0.17.2          | 916 KB    | : 100% 1.0/1 [00:00<00:00,  1.21it/s]\n",
            "numpy-1.19.1         | 21 KB     | : 100% 1.0/1 [00:00<00:00,  7.99it/s]\n",
            "joblib-0.16.0        | 210 KB    | : 100% 1.0/1 [00:00<00:00,  5.71it/s]\n",
            "libffi-3.3           | 50 KB     | : 100% 1.0/1 [00:00<00:00,  7.17it/s]\n",
            "libstdcxx-ng-9.1.0   | 3.1 MB    | : 100% 1.0/1 [00:00<00:00,  3.72it/s]\n",
            "numpy-base-1.19.1    | 4.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.55it/s]\n",
            "libsodium-1.0.18     | 244 KB    | : 100% 1.0/1 [00:00<00:00,  6.86it/s]\n",
            "mkl_random-1.1.1     | 327 KB    | : 100% 1.0/1 [00:00<00:00,  6.38it/s]\n",
            "fontconfig-2.13.0    | 227 KB    | : 100% 1.0/1 [00:00<00:00,  7.01it/s]\n",
            "pytz-2020.1          | 184 KB    | : 100% 1.0/1 [00:00<00:00,  4.99it/s]\n",
            "traitlets-4.3.3      | 140 KB    | : 100% 1.0/1 [00:00<00:00,  6.75it/s]\n",
            "wcwidth-0.2.5        | 29 KB     | : 100% 1.0/1 [00:00<00:00,  7.26it/s]\n",
            "ca-certificates-2020 | 125 KB    | : 100% 1.0/1 [00:00<00:00,  7.42it/s]\n",
            "cycler-0.10.0        | 13 KB     | : 100% 1.0/1 [00:00<00:00,  6.68it/s]\n",
            "matplotlib-3.1.1     | 5.0 MB    | : 100% 1.0/1 [00:00<00:00,  2.65it/s]\n",
            "decorator-4.4.2      | 14 KB     | : 100% 1.0/1 [00:00<00:00,  8.63it/s]\n",
            "qt-5.9.7             | 68.5 MB   | : 100% 1.0/1 [00:04<00:00,  4.26s/it]\n",
            "libxml2-2.9.10       | 1.2 MB    | : 100% 1.0/1 [00:00<00:00,  4.15it/s]\n",
            "icu-58.2             | 10.5 MB   | : 100% 1.0/1 [00:00<00:00,  1.60it/s]\n",
            "blas-1.0             | 6 KB      | : 100% 1.0/1 [00:00<00:00,  8.14it/s]\n",
            "tk-8.6.10            | 3.0 MB    | : 100% 1.0/1 [00:00<00:00,  3.61it/s]\n",
            "zeromq-4.3.2         | 511 KB    | : 100% 1.0/1 [00:00<00:00,  6.04it/s]\n",
            "ipython-7.16.1       | 999 KB    | : 100% 1.0/1 [00:00<00:00,  3.95it/s]\n",
            "scikit-learn-0.21.3  | 5.0 MB    | : 100% 1.0/1 [00:00<00:00,  2.17it/s]               \n",
            "jupyter_core-4.6.3   | 71 KB     | : 100% 1.0/1 [00:00<00:00,  5.98it/s]\n",
            "pyzmq-19.0.1         | 460 KB    | : 100% 1.0/1 [00:04<00:00,  4.56s/it]\n",
            "ld_impl_linux-64-2.3 | 568 KB    | : 100% 1.0/1 [00:00<00:00,  6.49it/s]\n",
            "pip-20.2.2           | 1.8 MB    | : 100% 1.0/1 [00:00<00:00,  3.19it/s]\n",
            "kiwisolver-1.2.0     | 84 KB     | : 100% 1.0/1 [00:00<00:00,  6.93it/s]\n",
            "pygments-2.6.1       | 654 KB    | : 100% 1.0/1 [00:00<00:00,  4.96it/s]\n",
            "expat-2.2.9          | 156 KB    | : 100% 1.0/1 [00:00<00:00,  7.56it/s]\n",
            "jupyter_client-6.1.6 | 84 KB     | : 100% 1.0/1 [00:00<00:00,  7.54it/s]\n",
            "gst-plugins-base-1.1 | 4.8 MB    | : 100% 1.0/1 [00:00<00:00,  3.22it/s]\n",
            "libxcb-1.14          | 505 KB    | : 100% 1.0/1 [00:00<00:00,  6.44it/s]\n",
            "ipython_genutils-0.2 | 39 KB     | : 100% 1.0/1 [00:00<00:00,  6.98it/s]\n",
            "pcre-8.44            | 212 KB    | : 100% 1.0/1 [00:00<00:00,  7.62it/s]\n",
            "setuptools-49.6.0    | 757 KB    | : 100% 1.0/1 [00:00<00:00,  5.01it/s]\n",
            "pyqt-5.9.2           | 4.5 MB    | : 100% 1.0/1 [00:00<00:00,  2.03it/s]\n",
            "libpng-1.6.37        | 278 KB    | : 100% 1.0/1 [00:00<00:00,  6.77it/s]\n",
            "pyparsing-2.4.7      | 65 KB     | : 100% 1.0/1 [00:00<00:00,  7.66it/s]\n",
            "python-dateutil-2.8. | 215 KB    | : 100% 1.0/1 [00:00<00:00,  7.53it/s]\n",
            "cffi-1.14.2          | 227 KB    | : 100% 1.0/1 [00:00<00:00,  6.54it/s]\n",
            "python-3.6.12        | 29.7 MB   | : 100% 1.0/1 [00:01<00:00,  1.50s/it]              \n",
            "ninja-1.10.1         | 1.4 MB    | : 100% 1.0/1 [00:00<00:00,  4.43it/s]               \n",
            "six-1.15.0           | 13 KB     | : 100% 1.0/1 [00:00<00:00,  8.57it/s]\n",
            "parso-0.7.0          | 72 KB     | : 100% 1.0/1 [00:00<00:00,  7.97it/s]\n",
            "freetype-2.10.2      | 608 KB    | : 100% 1.0/1 [00:00<00:00,  6.25it/s]\n",
            "tornado-6.0.4        | 597 KB    | : 100% 1.0/1 [00:00<00:00,  5.72it/s]\n",
            "dbus-1.13.16         | 501 KB    | : 100% 1.0/1 [00:00<00:00,  6.82it/s]\n",
            "ipykernel-5.3.4      | 181 KB    | : 100% 1.0/1 [00:00<00:00,  6.10it/s]\n",
            "ptyprocess-0.6.0     | 23 KB     | : 100% 1.0/1 [00:00<00:00,  7.89it/s]\n",
            "intel-openmp-2020.2  | 786 KB    | : 100% 1.0/1 [00:00<00:00,  6.68it/s]\n",
            "wheel-0.35.1         | 37 KB     | : 100% 1.0/1 [00:00<00:00,  7.74it/s]\n",
            "certifi-2020.6.20    | 156 KB    | : 100% 1.0/1 [00:00<00:00,  7.38it/s]\n",
            "pytorch-1.2.0        | 427.7 MB  | : 100% 1.0/1 [01:03<00:00, 63.50s/it]               \n",
            "libedit-3.1.20191231 | 116 KB    | : 100% 1.0/1 [00:00<00:00,  1.83it/s]\n",
            "libgfortran-ng-7.3.0 | 1006 KB   | : 100% 1.0/1 [00:00<00:00,  6.39it/s]\n",
            "openssl-1.1.1g       | 2.5 MB    | : 100% 1.0/1 [00:00<00:00,  3.08it/s]               \n",
            "mkl-service-2.3.0    | 219 KB    | : 100% 1.0/1 [00:00<00:00,  7.61it/s]\n",
            "_pytorch_select-0.1  | 3 KB      | : 100% 1.0/1 [00:00<00:00,  7.87it/s]\n",
            "backcall-0.2.0       | 15 KB     | : 100% 1.0/1 [00:00<00:00,  7.95it/s]\n",
            "sip-4.19.8           | 274 KB    | : 100% 1.0/1 [00:00<00:00,  7.58it/s]\n",
            "pycparser-2.20       | 94 KB     | : 100% 1.0/1 [00:00<00:00,  7.43it/s]\n",
            "readline-8.0         | 356 KB    | : 100% 1.0/1 [00:00<00:00,  7.48it/s]\n",
            "pexpect-4.8.0        | 82 KB     | : 100% 1.0/1 [00:00<00:00,  6.15it/s]\n",
            "scipy-1.5.2          | 14.4 MB   | : 100% 1.0/1 [00:01<00:00,  1.18s/it]\n",
            "mkl_fft-1.1.0        | 144 KB    | : 100% 1.0/1 [00:00<00:00,  8.79it/s]\n",
            "glib-2.65.0          | 2.9 MB    | : 100% 1.0/1 [00:00<00:00,  3.32it/s]\n",
            "mkl-2020.2           | 138.3 MB  | : 100% 1.0/1 [00:07<00:00,  7.85s/it]               \n",
            "prompt-toolkit-3.0.7 | 248 KB    | : 100% 1.0/1 [00:00<00:00,  7.04it/s]\n",
            "libuuid-1.0.3        | 15 KB     | : 100% 1.0/1 [00:00<00:00,  9.32it/s]\n",
            "gstreamer-1.14.0     | 3.1 MB    | : 100% 1.0/1 [00:00<00:00,  5.03it/s]\n",
            "Preparing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Verifying transaction: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Executing transaction: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Installing pip dependencies: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- Ran pip subprocess with arguments:\n",
            "['/usr/local/envs/deeptfactor/bin/python', '-m', 'pip', 'install', '-U', '-r', '/content/deeptfactor/condaenv.ayhbqbij.requirements.txt']\n",
            "Pip subprocess output:\n",
            "Collecting biopython==1.78\n",
            "  Downloading biopython-1.78-cp36-cp36m-manylinux1_x86_64.whl (2.3 MB)\n",
            "Requirement already up-to-date: torch==1.2.0 in /usr/local/envs/deeptfactor/lib/python3.6/site-packages (from -r /content/deeptfactor/condaenv.ayhbqbij.requirements.txt (line 2)) (1.2.0)\n",
            "Requirement already satisfied, skipping upgrade: numpy in /usr/local/envs/deeptfactor/lib/python3.6/site-packages (from biopython==1.78->-r /content/deeptfactor/condaenv.ayhbqbij.requirements.txt (line 1)) (1.19.1)\n",
            "Installing collected packages: biopython\n",
            "Successfully installed biopython-1.78\n",
            "\n",
            "\b\bdone\n",
            "#\n",
            "# To activate this environment, use\n",
            "#\n",
            "#     $ conda activate deeptfactor\n",
            "#\n",
            "# To deactivate an active environment, use\n",
            "#\n",
            "#     $ conda deactivate\n",
            "\n",
            "Collecting bipython\n",
            "  Downloading bipython-0.1.2.tar.gz (32 kB)\n",
            "Collecting Pygments>=1.6\n",
            "  Downloading Pygments-2.10.0-py3-none-any.whl (1.0 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.0 MB 8.2 MB/s \n",
            "\u001b[?25hCollecting urwid>=1.1.1\n",
            "  Downloading urwid-2.1.2.tar.gz (634 kB)\n",
            "\u001b[K     |████████████████████████████████| 634 kB 43.9 MB/s \n",
            "\u001b[?25hCollecting bpython>=0.12\n",
            "  Downloading bpython-0.22.1-py3-none-any.whl (187 kB)\n",
            "\u001b[K     |████████████████████████████████| 187 kB 47.4 MB/s \n",
            "\u001b[?25hCollecting pyzmq>=2.1.11\n",
            "  Downloading pyzmq-22.3.0-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (1.1 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.1 MB 41.2 MB/s \n",
            "\u001b[?25hCollecting ipython>=1.0\n",
            "  Downloading ipython-7.30.1-py3-none-any.whl (791 kB)\n",
            "\u001b[K     |████████████████████████████████| 791 kB 46.1 MB/s \n",
            "\u001b[?25hCollecting greenlet\n",
            "  Downloading greenlet-1.1.2-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (150 kB)\n",
            "\u001b[K     |████████████████████████████████| 150 kB 54.9 MB/s \n",
            "\u001b[?25hCollecting pyxdg\n",
            "  Downloading pyxdg-0.27-py2.py3-none-any.whl (49 kB)\n",
            "\u001b[K     |████████████████████████████████| 49 kB 6.2 MB/s \n",
            "\u001b[?25hCollecting backports.cached-property\n",
            "  Downloading backports.cached_property-1.0.1-py3-none-any.whl (5.7 kB)\n",
            "Collecting cwcwidth\n",
            "  Downloading cwcwidth-0.1.5-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (54 kB)\n",
            "\u001b[K     |████████████████████████████████| 54 kB 2.6 MB/s \n",
            "\u001b[?25hCollecting typing-extensions\n",
            "  Downloading typing_extensions-4.0.1-py3-none-any.whl (22 kB)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/site-packages (from bpython>=0.12->bipython) (2.26.0)\n",
            "Collecting curtsies>=0.3.5\n",
            "  Downloading curtsies-0.3.10.tar.gz (53 kB)\n",
            "\u001b[K     |████████████████████████████████| 53 kB 1.8 MB/s \n",
            "\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "    Preparing wheel metadata ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting blessings>=1.5\n",
            "  Downloading blessings-1.7-py3-none-any.whl (18 kB)\n",
            "Requirement already satisfied: six in /usr/local/lib/python3.7/site-packages (from blessings>=1.5->curtsies>=0.3.5->bpython>=0.12->bipython) (1.16.0)\n",
            "Requirement already satisfied: setuptools>=18.5 in /usr/local/lib/python3.7/site-packages (from ipython>=1.0->bipython) (58.0.4)\n",
            "Collecting decorator\n",
            "  Downloading decorator-5.1.0-py3-none-any.whl (9.1 kB)\n",
            "Collecting prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0\n",
            "  Downloading prompt_toolkit-3.0.24-py3-none-any.whl (374 kB)\n",
            "\u001b[K     |████████████████████████████████| 374 kB 55.3 MB/s \n",
            "\u001b[?25hCollecting matplotlib-inline\n",
            "  Downloading matplotlib_inline-0.1.3-py3-none-any.whl (8.2 kB)\n",
            "Collecting backcall\n",
            "  Downloading backcall-0.2.0-py2.py3-none-any.whl (11 kB)\n",
            "Collecting pickleshare\n",
            "  Downloading pickleshare-0.7.5-py2.py3-none-any.whl (6.9 kB)\n",
            "Collecting jedi>=0.16\n",
            "  Downloading jedi-0.18.1-py2.py3-none-any.whl (1.6 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.6 MB 52.3 MB/s \n",
            "\u001b[?25hCollecting traitlets>=4.2\n",
            "  Downloading traitlets-5.1.1-py3-none-any.whl (102 kB)\n",
            "\u001b[K     |████████████████████████████████| 102 kB 12.1 MB/s \n",
            "\u001b[?25hCollecting pexpect>4.3\n",
            "  Downloading pexpect-4.8.0-py2.py3-none-any.whl (59 kB)\n",
            "\u001b[K     |████████████████████████████████| 59 kB 6.4 MB/s \n",
            "\u001b[?25hCollecting parso<0.9.0,>=0.8.0\n",
            "  Downloading parso-0.8.3-py2.py3-none-any.whl (100 kB)\n",
            "\u001b[K     |████████████████████████████████| 100 kB 9.5 MB/s \n",
            "\u001b[?25hCollecting ptyprocess>=0.5\n",
            "  Downloading ptyprocess-0.7.0-py2.py3-none-any.whl (13 kB)\n",
            "Collecting wcwidth\n",
            "  Downloading wcwidth-0.2.5-py2.py3-none-any.whl (30 kB)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.7/site-packages (from requests->bpython>=0.12->bipython) (3.3)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/local/lib/python3.7/site-packages (from requests->bpython>=0.12->bipython) (1.26.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/site-packages (from requests->bpython>=0.12->bipython) (2021.10.8)\n",
            "Requirement already satisfied: charset-normalizer~=2.0.0 in /usr/local/lib/python3.7/site-packages (from requests->bpython>=0.12->bipython) (2.0.4)\n",
            "Building wheels for collected packages: bipython, curtsies, urwid\n",
            "  Building wheel for bipython (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for bipython: filename=bipython-0.1.2-py3-none-any.whl size=35075 sha256=f1fd1f87e8ec9b39846a2c89bc71793ceeb0b21cc2d408d3e56d075f1a924b8b\n",
            "  Stored in directory: /root/.cache/pip/wheels/7b/69/23/6a579fd64f8939757e242fa9be997f5ab887adde79a10f2b37\n",
            "  Building wheel for curtsies (PEP 517) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for curtsies: filename=curtsies-0.3.10-py3-none-any.whl size=35214 sha256=35054708331c4e4d4251a3c794ccbea7d4f2e53f8022bf5e1c2c845e7a379840\n",
            "  Stored in directory: /root/.cache/pip/wheels/0a/68/71/bdac55722d0d1a909e2ea519f4f908e07759a607c7b15899da\n",
            "  Building wheel for urwid (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for urwid: filename=urwid-2.1.2-cp37-cp37m-linux_x86_64.whl size=260315 sha256=02d670e0857789df0d67eebf2a3e46fca3a58c9dd10ea6603c89937ffd38a5b9\n",
            "  Stored in directory: /root/.cache/pip/wheels/79/77/cf/cae9cf1cc3f1f777f9db531424bbd9e15aa38e4ca28dbe499e\n",
            "Successfully built bipython curtsies urwid\n",
            "Installing collected packages: wcwidth, traitlets, ptyprocess, parso, cwcwidth, blessings, backports.cached-property, typing-extensions, pyxdg, Pygments, prompt-toolkit, pickleshare, pexpect, matplotlib-inline, jedi, greenlet, decorator, curtsies, backcall, urwid, pyzmq, ipython, bpython, bipython\n",
            "Successfully installed Pygments-2.10.0 backcall-0.2.0 backports.cached-property-1.0.1 bipython-0.1.2 blessings-1.7 bpython-0.22.1 curtsies-0.3.10 cwcwidth-0.1.5 decorator-5.1.0 greenlet-1.1.2 ipython-7.30.1 jedi-0.18.1 matplotlib-inline-0.1.3 parso-0.8.3 pexpect-4.8.0 pickleshare-0.7.5 prompt-toolkit-3.0.24 ptyprocess-0.7.0 pyxdg-0.27 pyzmq-22.3.0 traitlets-5.1.1 typing-extensions-4.0.1 urwid-2.1.2 wcwidth-0.2.5\n",
            "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "pexpect",
                  "pickleshare",
                  "traitlets",
                  "wcwidth",
                  "zmq"
                ]
              }
            }
          },
          "metadata": {}
        }
      ],
      "source": [
        "!wget -c https://repo.continuum.io/archive/Anaconda3-5.1.0-Linux-x86_64.sh\n",
        "\n",
        "!chmod +x Anaconda3-5.1.0-Linux-x86_64.sh\n",
        "\n",
        "!bash ./Anaconda3-5.1.0-Linux-x86_64.sh -b -f -p /usr/local\n",
        "!conda update conda -y -q\n",
        "!source /usr/local/etc/profile.d/conda.sh\n",
        "!conda init \n",
        "\n",
        "!git clone https://bitbucket.org/kaistsystemsbiology/deeptfactor.git\n",
        "\n",
        "%cd deeptfactor\n",
        "\n",
        "!conda env create -f environment.yml\n",
        "\n",
        "!source activate deeptfactor\n",
        "!pip install bipython"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## 2. Load software packages"
      ],
      "metadata": {
        "id": "TEU5j19aED0S"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import random\n",
        "import numpy as np\n",
        "\n",
        "import torch\n",
        "import torch.nn as nn\n",
        "import torch.optim as optim\n",
        "from torch.utils.data import DataLoader\n",
        "from ipywidgets import interact\n",
        "import ipywidgets as widgets\n",
        "\n",
        "from deeptfactor.data_loader import EnzymeDataset\n",
        "from deeptfactor.models import DeepTFactor\n",
        "\n",
        "class Data:\n",
        "    def __init__(self):\n",
        "        self.results = []"
      ],
      "metadata": {
        "id": "V_00YhtC1Ksg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## 3. Make predictions with pretrained DeepTFactor model\n",
        "\n",
        "Below, you can introduce a FASTA file with the sequences of all the proteins we want the model to predict as factor transcriptions or not factor transcriptions. As an example, I have introduced a FASTA file with all reviewed human proteins in UniProt associated with the term \"pain\":"
      ],
      "metadata": {
        "id": "SxmYIEIhEHv2"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "data = Data()\n",
        "\n",
        "fasta = widgets.FileUpload(description = \"Introduce FASTA file\",\n",
        "                           style={'description_width': 'initial'})\n",
        "\n",
        "#@title\n",
        "device = \"cpu\"\n",
        "num_cpu = 1\n",
        "batch_size = 128\n",
        "\n",
        "\n",
        "checkpt_file = \"./trained_model/DeepTFactor_ckpt.pt\"\n",
        "\n",
        "@interact\n",
        "def predict_deeptfactor(fasta = fasta):\n",
        "    if not fasta:\n",
        "        return None\n",
        "      \n",
        "    torch.set_num_threads(num_cpu)\n",
        "\n",
        "    d = list(fasta.values())[0][\"content\"].decode(\"utf-8\")\n",
        "    seqs = \"\".join([l if not l.startswith(\">\") else \">\"\n",
        "                    for l in d.splitlines()]).strip(\">\").split(\">\")\n",
        "    protein_seqs = [s if len(s)==1000 else s[:1000] + \"_\"*(1000-len(s))\n",
        "                    for s in seqs]\n",
        "\n",
        "    seq_ids = [l for l in d.splitlines() if l.startswith(\">\")]\n",
        "\n",
        "    pseudo_labels = np.zeros((len(protein_seqs)))\n",
        "    proteinDataset = EnzymeDataset(protein_seqs, pseudo_labels)\n",
        "    proteinDataloader = DataLoader(proteinDataset,\n",
        "                                   batch_size=batch_size,\n",
        "                                   shuffle=False)\n",
        "\n",
        "\n",
        "    model = DeepTFactor(out_features=[1])\n",
        "    model = model.to(device)\n",
        "\n",
        "    ckpt = torch.load(f'{checkpt_file}', map_location=device)\n",
        "    model.load_state_dict(ckpt['model'])\n",
        "    cutoff = 0.5\n",
        "\n",
        "    y_pred = torch.zeros([len(seq_ids), 1])\n",
        "    with torch.no_grad():\n",
        "        model.eval()\n",
        "        cnt = 0\n",
        "        for x, _ in proteinDataloader:\n",
        "            x = x.type(torch.FloatTensor)\n",
        "            x_length = x.shape[0]\n",
        "            output = model(x.to(device))\n",
        "            prediction = output.cpu()\n",
        "            y_pred[cnt:cnt+x_length] = prediction\n",
        "            cnt += x_length\n",
        "\n",
        "    scores = y_pred[:,0]\n",
        "\n",
        "    for seq_id, score in zip(seq_ids, scores):\n",
        "        if score > cutoff:\n",
        "            tf = True\n",
        "        else:\n",
        "            tf = False\n",
        "        data.results.append({\"seq_id\": seq_id,\n",
        "                             \"score\": score.numpy(),\n",
        "                             \"is_tf\": tf})"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 49,
          "referenced_widgets": [
            "380286e962254235b4a49ca3becb9683",
            "11644cd5bb8a4b86a90c3ba7fca2a09d",
            "00a9610e16fc49c0905d06888c44f6ec",
            "de9f5e391d154db1b7f1129225b39cc0",
            "3b680612e5e2453ca5bf42a8d7499ccf",
            "2282b02b0e5b4c9c8e49938586b0eb78",
            "a9adc18bdd27416498bf9cebe159c9de"
          ]
        },
        "id": "8BYPtRom2Y2d",
        "outputId": "4a86b036-5f2a-4e19-8978-e976135c35a2"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.jupyter.widget-view+json": {
              "model_id": "380286e962254235b4a49ca3becb9683",
              "version_minor": 0,
              "version_major": 2
            },
            "text/plain": [
              "interactive(children=(FileUpload(value={}, description='Introduce FASTA file'), Output()), _dom_classes=('widg…"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## 4. Analyze results\n",
        "\n",
        "After a while, we can analyze our results. If we represent them in the form of a table:"
      ],
      "metadata": {
        "id": "LHa4Re8OERZd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "df = pd.DataFrame(data.results)\n",
        "df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "J7LJSGmUDJMd",
        "outputId": "62f6a8c1-5d81-4b2d-aee3-855eab496017"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>seq_id</th>\n",
              "      <th>score</th>\n",
              "      <th>is_tf</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>&gt;sp|Q15858|SCN9A_HUMAN Sodium channel protein ...</td>\n",
              "      <td>0.008337229</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>&gt;sp|O75762|TRPA1_HUMAN Transient receptor pote...</td>\n",
              "      <td>0.0050662435</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>&gt;sp|P04629|NTRK1_HUMAN High affinity nerve gro...</td>\n",
              "      <td>0.010042973</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>&gt;sp|Q9UI33|SCNBA_HUMAN Sodium channel protein ...</td>\n",
              "      <td>0.008308047</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>&gt;sp|Q9Y5Y9|SCNAA_HUMAN Sodium channel protein ...</td>\n",
              "      <td>0.0052825967</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>283</th>\n",
              "      <td>&gt;sp|P18509|PACA_HUMAN Pituitary adenylate cycl...</td>\n",
              "      <td>0.01865728</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>284</th>\n",
              "      <td>&gt;sp|Q96LR4|TAFA4_HUMAN Chemokine-like protein ...</td>\n",
              "      <td>0.007869436</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>285</th>\n",
              "      <td>&gt;sp|Q9BYJ1|LOXE3_HUMAN Hydroperoxide isomerase...</td>\n",
              "      <td>0.005664655</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>286</th>\n",
              "      <td>&gt;sp|P01210|PENK_HUMAN Proenkephalin-A OS=Homo ...</td>\n",
              "      <td>0.008034096</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>287</th>\n",
              "      <td>&gt;sp|Q9UPV7|PHF24_HUMAN PHD finger protein 24 O...</td>\n",
              "      <td>0.09844679</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>288 rows × 3 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                seq_id         score  is_tf\n",
              "0    >sp|Q15858|SCN9A_HUMAN Sodium channel protein ...   0.008337229  False\n",
              "1    >sp|O75762|TRPA1_HUMAN Transient receptor pote...  0.0050662435  False\n",
              "2    >sp|P04629|NTRK1_HUMAN High affinity nerve gro...   0.010042973  False\n",
              "3    >sp|Q9UI33|SCNBA_HUMAN Sodium channel protein ...   0.008308047  False\n",
              "4    >sp|Q9Y5Y9|SCNAA_HUMAN Sodium channel protein ...  0.0052825967  False\n",
              "..                                                 ...           ...    ...\n",
              "283  >sp|P18509|PACA_HUMAN Pituitary adenylate cycl...    0.01865728  False\n",
              "284  >sp|Q96LR4|TAFA4_HUMAN Chemokine-like protein ...   0.007869436  False\n",
              "285  >sp|Q9BYJ1|LOXE3_HUMAN Hydroperoxide isomerase...   0.005664655  False\n",
              "286  >sp|P01210|PENK_HUMAN Proenkephalin-A OS=Homo ...   0.008034096  False\n",
              "287  >sp|Q9UPV7|PHF24_HUMAN PHD finger protein 24 O...    0.09844679  False\n",
              "\n",
              "[288 rows x 3 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "The majority of proteins have been predicted as not transcription factors:"
      ],
      "metadata": {
        "id": "c-3hqzvmpAAz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "print(df.is_tf.mean().round(3)*100, \"% of proteins have been predicted as transcription factors.\", sep = \"\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yydL5WgYoxl0",
        "outputId": "d1e08066-3987-46a0-ae61-27f1fe17101c"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "5.2% of proteins have been predicted as transcription factors.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We can complement our analysis with some data from UniProt. For each protein, we are downloading its function description:"
      ],
      "metadata": {
        "id": "ASyoEzcEpMxE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import requests\n",
        "from bs4 import BeautifulSoup as bs\n",
        "\n",
        "df[\"uniprot_id\"] = [v.split(\"|\")[1] for v in df.seq_id.values.tolist()]\n",
        "\n",
        "def get_descr(up_id):\n",
        "  page = requests.get(\"https://www.uniprot.org/uniprot/\" + up_id).text\n",
        "  descr = bs(page, \"html.parser\").find(\"div\", {\"id\": \"function\"}).p.text\n",
        "  return descr\n",
        "\n",
        "df[\"uniprot_description\"] = [get_descr(r.uniprot_id) for i,r in df.iterrows()]\n",
        "df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "GEl2IjQpH3ri",
        "outputId": "2af5efa9-7999-403f-85d1-06943af89807"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>seq_id</th>\n",
              "      <th>score</th>\n",
              "      <th>is_tf</th>\n",
              "      <th>uniprot_id</th>\n",
              "      <th>uniprot_description</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>&gt;sp|Q15858|SCN9A_HUMAN Sodium channel protein ...</td>\n",
              "      <td>0.008337229</td>\n",
              "      <td>False</td>\n",
              "      <td>Q15858</td>\n",
              "      <td>Mediates the voltage-dependent sodium ion perm...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>&gt;sp|O75762|TRPA1_HUMAN Transient receptor pote...</td>\n",
              "      <td>0.0050662435</td>\n",
              "      <td>False</td>\n",
              "      <td>O75762</td>\n",
              "      <td>Receptor-activated non-selective cation channe...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>&gt;sp|P04629|NTRK1_HUMAN High affinity nerve gro...</td>\n",
              "      <td>0.010042973</td>\n",
              "      <td>False</td>\n",
              "      <td>P04629</td>\n",
              "      <td>Receptor tyrosine kinase involved in the devel...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>&gt;sp|Q9UI33|SCNBA_HUMAN Sodium channel protein ...</td>\n",
              "      <td>0.008308047</td>\n",
              "      <td>False</td>\n",
              "      <td>Q9UI33</td>\n",
              "      <td>This protein mediates the voltage-dependent so...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>&gt;sp|Q9Y5Y9|SCNAA_HUMAN Sodium channel protein ...</td>\n",
              "      <td>0.0052825967</td>\n",
              "      <td>False</td>\n",
              "      <td>Q9Y5Y9</td>\n",
              "      <td>Tetrodotoxin-resistant channel that mediates t...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>283</th>\n",
              "      <td>&gt;sp|P18509|PACA_HUMAN Pituitary adenylate cycl...</td>\n",
              "      <td>0.01865728</td>\n",
              "      <td>False</td>\n",
              "      <td>P18509</td>\n",
              "      <td>Binding to its receptor activates G proteins a...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>284</th>\n",
              "      <td>&gt;sp|Q96LR4|TAFA4_HUMAN Chemokine-like protein ...</td>\n",
              "      <td>0.007869436</td>\n",
              "      <td>False</td>\n",
              "      <td>Q96LR4</td>\n",
              "      <td>Modulates injury-induced and chemical pain hyp...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>285</th>\n",
              "      <td>&gt;sp|Q9BYJ1|LOXE3_HUMAN Hydroperoxide isomerase...</td>\n",
              "      <td>0.005664655</td>\n",
              "      <td>False</td>\n",
              "      <td>Q9BYJ1</td>\n",
              "      <td>Non-heme iron-containing lipoxygenase which is...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>286</th>\n",
              "      <td>&gt;sp|P01210|PENK_HUMAN Proenkephalin-A OS=Homo ...</td>\n",
              "      <td>0.008034096</td>\n",
              "      <td>False</td>\n",
              "      <td>P01210</td>\n",
              "      <td>Competes with and mimic the effects of opiate ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>287</th>\n",
              "      <td>&gt;sp|Q9UPV7|PHF24_HUMAN PHD finger protein 24 O...</td>\n",
              "      <td>0.09844679</td>\n",
              "      <td>False</td>\n",
              "      <td>Q9UPV7</td>\n",
              "      <td>Pathway Commons web resource for biological pa...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>288 rows × 5 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                seq_id  ...                                uniprot_description\n",
              "0    >sp|Q15858|SCN9A_HUMAN Sodium channel protein ...  ...  Mediates the voltage-dependent sodium ion perm...\n",
              "1    >sp|O75762|TRPA1_HUMAN Transient receptor pote...  ...  Receptor-activated non-selective cation channe...\n",
              "2    >sp|P04629|NTRK1_HUMAN High affinity nerve gro...  ...  Receptor tyrosine kinase involved in the devel...\n",
              "3    >sp|Q9UI33|SCNBA_HUMAN Sodium channel protein ...  ...  This protein mediates the voltage-dependent so...\n",
              "4    >sp|Q9Y5Y9|SCNAA_HUMAN Sodium channel protein ...  ...  Tetrodotoxin-resistant channel that mediates t...\n",
              "..                                                 ...  ...                                                ...\n",
              "283  >sp|P18509|PACA_HUMAN Pituitary adenylate cycl...  ...  Binding to its receptor activates G proteins a...\n",
              "284  >sp|Q96LR4|TAFA4_HUMAN Chemokine-like protein ...  ...  Modulates injury-induced and chemical pain hyp...\n",
              "285  >sp|Q9BYJ1|LOXE3_HUMAN Hydroperoxide isomerase...  ...  Non-heme iron-containing lipoxygenase which is...\n",
              "286  >sp|P01210|PENK_HUMAN Proenkephalin-A OS=Homo ...  ...  Competes with and mimic the effects of opiate ...\n",
              "287  >sp|Q9UPV7|PHF24_HUMAN PHD finger protein 24 O...  ...  Pathway Commons web resource for biological pa...\n",
              "\n",
              "[288 rows x 5 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "For example, if we filter out all the proteins that have not been predicted as transcription factors:"
      ],
      "metadata": {
        "id": "BNruaIO2pitB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "df[df.is_tf]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 520
        },
        "id": "oplgfJ91iUw9",
        "outputId": "c332d384-3305-49a2-db4a-b0eeb2b4e17e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>seq_id</th>\n",
              "      <th>score</th>\n",
              "      <th>is_tf</th>\n",
              "      <th>uniprot_id</th>\n",
              "      <th>uniprot_description</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>&gt;sp|Q9H4Q4|PRD12_HUMAN PR domain zinc finger p...</td>\n",
              "      <td>0.9780904</td>\n",
              "      <td>True</td>\n",
              "      <td>Q9H4Q4</td>\n",
              "      <td>Involved in the positive regulation of histone...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td>&gt;sp|P37275|ZEB1_HUMAN Zinc finger E-box-bindin...</td>\n",
              "      <td>0.994453</td>\n",
              "      <td>True</td>\n",
              "      <td>P37275</td>\n",
              "      <td>Acts as a transcriptional repressor. Inhibits ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>56</th>\n",
              "      <td>&gt;sp|Q8N1G0|ZN687_HUMAN Zinc finger protein 687...</td>\n",
              "      <td>0.8928037</td>\n",
              "      <td>True</td>\n",
              "      <td>Q8N1G0</td>\n",
              "      <td>May be involved in transcriptional regulation.</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>110</th>\n",
              "      <td>&gt;sp|P78543|BTG2_HUMAN Protein BTG2 OS=Homo sap...</td>\n",
              "      <td>0.98999715</td>\n",
              "      <td>True</td>\n",
              "      <td>P78543</td>\n",
              "      <td>Anti-proliferative protein; the function is me...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>135</th>\n",
              "      <td>&gt;sp|A6NNA5|DRGX_HUMAN Dorsal root ganglia home...</td>\n",
              "      <td>0.99419653</td>\n",
              "      <td>True</td>\n",
              "      <td>A6NNA5</td>\n",
              "      <td>Transcription factor required for the formatio...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>139</th>\n",
              "      <td>&gt;sp|Q15306|IRF4_HUMAN Interferon regulatory fa...</td>\n",
              "      <td>0.9757959</td>\n",
              "      <td>True</td>\n",
              "      <td>Q15306</td>\n",
              "      <td>Transcriptional activator. Binds to the interf...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>148</th>\n",
              "      <td>&gt;sp|P15884|ITF2_HUMAN Transcription factor 4 O...</td>\n",
              "      <td>0.9588655</td>\n",
              "      <td>True</td>\n",
              "      <td>P15884</td>\n",
              "      <td>Transcription factor that binds to the immunog...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>198</th>\n",
              "      <td>&gt;sp|Q9BR39|JPH2_HUMAN Junctophilin-2 OS=Homo s...</td>\n",
              "      <td>0.97406197</td>\n",
              "      <td>True</td>\n",
              "      <td>Q9BR39</td>\n",
              "      <td>Membrane-binding protein that provides a struc...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>204</th>\n",
              "      <td>&gt;sp|O75398|DEAF1_HUMAN Deformed epidermal auto...</td>\n",
              "      <td>0.9718963</td>\n",
              "      <td>True</td>\n",
              "      <td>O75398</td>\n",
              "      <td>Transcription factor that binds to sequence wi...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>210</th>\n",
              "      <td>&gt;sp|P10588|NR2F6_HUMAN Nuclear receptor subfam...</td>\n",
              "      <td>0.9596225</td>\n",
              "      <td>True</td>\n",
              "      <td>P10588</td>\n",
              "      <td>Transcription factor predominantly involved in...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>216</th>\n",
              "      <td>&gt;sp|Q9GZZ0|HXD1_HUMAN Homeobox protein Hox-D1 ...</td>\n",
              "      <td>0.8528684</td>\n",
              "      <td>True</td>\n",
              "      <td>Q9GZZ0</td>\n",
              "      <td>Sequence-specific transcription factor which i...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>222</th>\n",
              "      <td>&gt;sp|P51608|MECP2_HUMAN Methyl-CpG-binding prot...</td>\n",
              "      <td>0.9455978</td>\n",
              "      <td>True</td>\n",
              "      <td>P51608</td>\n",
              "      <td>Chromosomal protein that binds to methylated D...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>242</th>\n",
              "      <td>&gt;sp|P61978|HNRPK_HUMAN Heterogeneous nuclear r...</td>\n",
              "      <td>0.9950328</td>\n",
              "      <td>True</td>\n",
              "      <td>P61978</td>\n",
              "      <td>One of the major pre-mRNA-binding proteins. Bi...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>266</th>\n",
              "      <td>&gt;sp|P17481|HXB8_HUMAN Homeobox protein Hox-B8 ...</td>\n",
              "      <td>0.95236325</td>\n",
              "      <td>True</td>\n",
              "      <td>P17481</td>\n",
              "      <td>Sequence-specific transcription factor which i...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>281</th>\n",
              "      <td>&gt;sp|Q99608|NECD_HUMAN Necdin OS=Homo sapiens O...</td>\n",
              "      <td>0.6428799</td>\n",
              "      <td>True</td>\n",
              "      <td>Q99608</td>\n",
              "      <td>Growth suppressor that facilitates the entry o...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                seq_id  ...                                uniprot_description\n",
              "15   >sp|Q9H4Q4|PRD12_HUMAN PR domain zinc finger p...  ...  Involved in the positive regulation of histone...\n",
              "48   >sp|P37275|ZEB1_HUMAN Zinc finger E-box-bindin...  ...  Acts as a transcriptional repressor. Inhibits ...\n",
              "56   >sp|Q8N1G0|ZN687_HUMAN Zinc finger protein 687...  ...     May be involved in transcriptional regulation.\n",
              "110  >sp|P78543|BTG2_HUMAN Protein BTG2 OS=Homo sap...  ...  Anti-proliferative protein; the function is me...\n",
              "135  >sp|A6NNA5|DRGX_HUMAN Dorsal root ganglia home...  ...  Transcription factor required for the formatio...\n",
              "139  >sp|Q15306|IRF4_HUMAN Interferon regulatory fa...  ...  Transcriptional activator. Binds to the interf...\n",
              "148  >sp|P15884|ITF2_HUMAN Transcription factor 4 O...  ...  Transcription factor that binds to the immunog...\n",
              "198  >sp|Q9BR39|JPH2_HUMAN Junctophilin-2 OS=Homo s...  ...  Membrane-binding protein that provides a struc...\n",
              "204  >sp|O75398|DEAF1_HUMAN Deformed epidermal auto...  ...  Transcription factor that binds to sequence wi...\n",
              "210  >sp|P10588|NR2F6_HUMAN Nuclear receptor subfam...  ...  Transcription factor predominantly involved in...\n",
              "216  >sp|Q9GZZ0|HXD1_HUMAN Homeobox protein Hox-D1 ...  ...  Sequence-specific transcription factor which i...\n",
              "222  >sp|P51608|MECP2_HUMAN Methyl-CpG-binding prot...  ...  Chromosomal protein that binds to methylated D...\n",
              "242  >sp|P61978|HNRPK_HUMAN Heterogeneous nuclear r...  ...  One of the major pre-mRNA-binding proteins. Bi...\n",
              "266  >sp|P17481|HXB8_HUMAN Homeobox protein Hox-B8 ...  ...  Sequence-specific transcription factor which i...\n",
              "281  >sp|Q99608|NECD_HUMAN Necdin OS=Homo sapiens O...  ...  Growth suppressor that facilitates the entry o...\n",
              "\n",
              "[15 rows x 5 columns]"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We see that in most of the predicted transcription factors, UniProt function description contains \"Transcription factor\" or similar terms."
      ],
      "metadata": {
        "id": "APyo4ixdp3qR"
      }
    }
  ]
}